backend/src/apiserver/storage/pipeline_store.go

```diff
@@ -84,6 +85,36 @@ type PipelineStore struct {
 	uuid util.UUIDGeneratorInterface
 }
 
+func (s *PipelineStore) GetPipelineByNameAndNamespace(name string, namespace string) (*model.Pipeline, error) {
+	sql, args, err := sq.
```

Comment:
Notice that I kept the name of the rpc to `GetPipelineByName`, but changed the others (like this one) to be GetPipelineByNameAndNamespece @zijianjoy WDYT?

Reply:
The naming is okay, but looks like you still call `GetPipelineByNameAndNamespace` even if namespace is empty (standalone mode). How about calling `GetPipeline` when it is standalone mode?

Reply:
`func (r *ResourceManager) GetPipeline(pipelineId string) (*model.Pipeline, error) {`
I do not know the pipelineId, but the name

Reply:
Oh you are right. Then can we construct the sql query based on the existence of namespace?

That means, only add `sq.Eq{"pipelines.Namespace": namespace},` when namespace is not empty string.

Reply:
Only the combination - name and namespace is unique. Even if the namespace is an empty string, we still need to provide it, otherwise, we could accidentally read a Pipeline that belongs to another namespace.

Reply:
If namespace is not specified during the request, you will create a SQL query with `sq.Eq{"pipelines.Namespace": ""}` condition in the `WHERE` clause, that means the namespace column needs to be an empty string to be selected. Will it find the right pipeline in the KFP standalone scenario? Can we write a unit test about it?

Reply:
Yes, it will find the right pipeline. 
- The default value of the namespace is an empty string no matter the installation.
https://github.com/kubeflow/pipelines/blob/fdd42ff3847b4d0bf2fe50a31cec4a81a307a69d/backend/src/apiserver/model/pipeline.go#L42

  `mysql> show create table pipelines;`
  
  >| pipelines | CREATE TABLE `pipelines` (
    `UUID` varchar(255) NOT NULL,
    `CreatedAtInSec` bigint(20) NOT NULL,
    `Name` varchar(255) NOT NULL,
    `Description` longtext NOT NULL,
    `Parameters` longtext NOT NULL,
    `Status` varchar(255) NOT NULL,
    `DefaultVersionId` varchar(255) DEFAULT NULL,
    `Namespace` varchar(63) DEFAULT '',
    PRIMARY KEY (`UUID`),
    UNIQUE KEY `name_namespace_index` (`Name`,`Namespace`)
  ) ENGINE=InnoDB DEFAULT CHARSET=latin1


- **CreatePipeline** func uses `GetNamespaceFromAPIResourceReferences` to find out the namespace. It initializes the namespace to empty string no matter if `resource_reference` is provided or not. https://github.com/kubeflow/pipelines/blob/fdd42ff3847b4d0bf2fe50a31cec4a81a307a69d/backend/src/apiserver/server/pipeline_server.go#L101
         https://github.com/kubeflow/pipelines/blob/fdd42ff3847b4d0bf2fe50a31cec4a81a307a69d/backend/src/apiserver/common/util.go#L28-L36
- **UploadPipeline** also defaults the namespaces to empty string (if `NamespaceStringQuery` is not in the URL)
 https://github.com/kubeflow/pipelines/blob/fdd42ff3847b4d0bf2fe50a31cec4a81a307a69d/backend/src/apiserver/server/pipeline_upload_server.go#L95

See this comment - it looks it is still applicable
https://github.com/kubeflow/pipelines/issues/4197#issuecomment-721570237

Reply:
Got it, thank you Diana for the explanation and reference! It makes sense to me for using empty string for shared namespace.